### PR TITLE
[WIP] Dots with regions

### PR DESCRIPTION
**Pull request by sergpolly**
just trying to see the changes together

### CODE REVIEW

**Comment by agalitsyna** [comment]
Major improvements: 
- Flexible load of regions table (similar to recent snipper fix https://github.com/open2c/cooltools/commit/3d134bf8078275ecdd7bafcff7be59a7cd998149)
- Everything in dot finder is now fetched by region name, and not by chromosome. If the exact position of the region is needed, the region table or its elements are passed.
- `heatmap_tiles_generator_diag` now fetches the regions, not chromosomes
- `assign_regions` from snipping utility moved to `lib/common.py`. It is used before the clustering step to assign regions to  the filtered pixels.
- `clustering_step` now works with assigned regions and not simply chromosomes (note that only intra-region pixels are clustered)
- `test_call_dots_cli` added to the tests. Checks that cli runs, nothing more sophisticated.
- Unused parameters commented out with TODO note.
- To make the tests less confusing, "mock" test files for dotfinder renamed into "dotfinder_mock" files.

Accompanied with working draft PR of open2c_examples: https://github.com/open2c/open2c_examples/pull/8

File: cooltools/cli/call_dots.py [hunk 1]
```diff
@@ -108,60 +123,60 @@
...
-    default=False,
-)
+### Unused options commented out. TODO: remove them or implement
+# @click.option(
+#     "-s", "--output-scores",
+#     help="At the moment it is a redundant option that"
+#     " does nothing. Reserve it for a better dump"
+#     " of convolved scores.",
+#     type=str,
+#     required=False,
```
Comment (sergpolly) [inline cooltools/cli/call_dots.py:144]: let's remove them all - they are more technical/debugging options - that rely on a "single-pass" scenario - i.e. when we store results of the convolution on disk (abandoned for now at least)

File: cooltools/cli/call_dots.py [hunk 2]
```diff
@@ -174,19 +189,19 @@ def call_dots(
...
     dots_clustering_radius,
     verbose,
-    output_scores,
-    output_hists,
     output_calls,
-    score_dump_mode,
-    temp_dir,
-    no_delete_temp,
+    # output_scores,
+    # output_hists,
```
Comment (sergpolly) [inline cooltools/cli/call_dots.py:194]: remove them

File: cooltools/cli/call_dots.py [hunk 3]
```diff
@@ -197,60 +212,115 @@ def call_dots(
...
+        if regions_table.shape[1] == 4:
+            regions_table = regions_table.rename(
+                columns={0: "chrom", 1: "start", 2: "end", 3: "name"}
+            )
+            regions_table = bioframe.parse_regions(regions_table)
+        else:
+            regions_table = regions_table.rename(
+                columns={0: "chrom", 1: "start", 2: "end"}
+            )
+            regions_table["name"] = list(
```
Comment (sergpolly) [inline cooltools/cli/call_dots.py:281]: @agalitsyna I think `bioframe.parse_regions` does that automatically ... - if you give it a 3 column bed - i'll auto-generate UCSC names for it.

So I wonder if this whole `if regions_table.shape[1] == 4` is really needed - I think `bioframe.parse_regions` could handle both 3 and 4 columns BED-like dataframes - shall we try that instead ?

File: cooltools/cli/call_dots.py [hunk 4]
```diff
@@ -197,60 +212,115 @@ def call_dots(
...
-    )
-    expected_bins = get_exp_bins(expected, expected_chroms)
-    cool_bins = clr.bins()[:]["chrom"].isin(expected_chroms).sum()
-    if not (expected_bins == cool_bins):
+    try:
+        for region_name, group in expected.reset_index().groupby(region_column_name):
+            n_diags = group.shape[0]
+            region = regions_table.set_index("name").loc[region_name]
+            lo, hi = clr.extent(region)
+            assert n_diags == (hi - lo)
```
Comment (sergpolly) [inline cooltools/cli/call_dots.py:308]: super useful !

File: cooltools/dotfinder.py [hunk 5]
```diff
@@ -511,20 +511,19 @@ def heatmap_tiles_generator_diag(clr, chroms, pad_size, tile_size, band_to_cover
...
         chromosome name, row index of the tile,
-        column index of the tile (chrom, tilei, tilej).
+        column index of the tile (region_name, tilei, tilej).
 
     """
 
-    for chrom in chroms:
-        chr_start, chr_stop = clr.extent(chrom)
+    for chrom, start, end, region_name in regions.itertuples(index=False):
+        region_begin, region_end = clr.extent(f"{chrom}:{start}-{end}")
```
Comment (sergpolly) [inline cooltools/dotfinder.py:524]: why UCSC ? just a tuple should work `(chrom,start,end)`

File: cooltools/dotfinder.py [hunk 6]
```diff
@@ -1741,21 +1739,25 @@ def clustering_step(
...
 
     """
+    # Annotate regions, if needed:
+
+    scores_df = scores_df.copy()
+    if (
+        not "region" in scores_df.columns
+    ):  # If input scores are not annotated by regions:
+        scores_df["region"] = np.where(
+            scores_df["chr1"] == scores_df["chrom2"], scores_df["chrom1"], np.nan
```
Comment (sergpolly) [inline cooltools/dotfinder.py:1749]: @agalitsyna don't `"chr1"` - looks suspicious here ... should it be `chrom1` - for cis ?

Comment (sergpolly) [reply cooltools/dotfinder.py:1749]: also `np.where` signature looks different `np.where(condition, [x,y])` ...

File: cooltools/cli/call_dots.py (hunk 3 above)
Comment (sergpolly) [reply cooltools/cli/call_dots.py:281]: 3,4 - are needed as it turned out

Comment (sergpolly) [reply cooltools/cli/call_dots.py]: I removed `name` assignment - otherwise intact

File: cooltools/dotfinder.py (hunk 6 above)
Comment (sergpolly) [reply cooltools/dotfinder.py:1749]: Need to check that !

File: cooltools/dotfinder.py [hunk 7]
```diff
@@ -1741,21 +1439,25 @@ def clustering_step(
...
```
Comment (agalitsyna) [inline cooltools/dotfinder.py:1449]:
```suggestion
            scores_df["chrom1"] == scores_df["chrom2"], scores_df["chrom1"], np.nan
```

**Review by agalitsyna — CHANGES_REQUESTED** [review]
Comment on np.where in dotfinder resolved

File: cooltools/dotfinder.py (hunk 6 above)
Comment (agalitsyna) [reply cooltools/dotfinder.py]: np.where can work similarly to ternary operator on vectors, pretty cool: https://numpy.org/doc/stable/reference/generated/numpy.where.html

File: cooltools/dotfinder.py [hunk 8]
```diff
@@ -1771,18 +1473,28 @@ def clustering_step(
...
+    if len(pixel_clust_list) == 0:
+        if verbose:
+            print("No clusters found! Output will be empty")
+        empty_output = pd.DataFrame(
+            [],
+            columns=list(scores_df.columns)
+            + ["region1", "region2", "c_label", "c_size", "cstart1", "cstart2"],
+        )
+        return empty_output  # Empty dataframe with the same columns as anticipated
     pixel_clust_df = pd.concat(pixel_clust_list, ignore_index=False)
```
Comment (agalitsyna) [inline cooltools/dotfinder.py:1486]:
```suggestion
    pixel_clust_df = pd.concat(pixel_clust_list, ignore_index=False) # Concatenate the clustering results for different regions
```

File: cooltools/dotfinder.py [hunk 9]
```diff
@@ -1771,18 +1473,28 @@ def clustering_step(
...
-    #prevents scores_df categorical values (all chroms, including chrM)
-    df['chrom1'] = df['chrom1'].astype(str)
-    df['chrom2'] = df['chrom2'].astype(str)
+    # prevents scores_df categorical values (all chroms, including chrM)
+    df["region1"] = df["region"].astype(str)
+    df["region2"] = df["region"].astype(str)
     # report only centroids with highest Observed:
-    chrom_clust_group = df.groupby(["chrom1", "chrom2", "c_label"])
+    chrom_clust_group = df.groupby(["region1", "region2", "c_label"])
     centroids = df.loc[chrom_clust_group[obs_raw_name].idxmax()]
```
Comment (agalitsyna) [inline cooltools/dotfinder.py:1498]:
```suggestion
    centroids = df.loc[chrom_clust_group[obs_raw_name].idxmax()] # Select the brightest pixel in the cluster
```

File: cooltools/dotfinder.py [hunk 10]
```diff
@@ -1741,21 +1439,25 @@ def clustering_step(
...
+    ):  # If input scores are not annotated by regions:
+        scores_df["region"] = np.where(
+            scores_df["chrom1"] == scores_df["chrom2"], scores_df["chrom1"], np.nan
+        )
     # using different bin12_id_names since all
     # pixels are annotated at this point.
     pixel_clust_list = []
-    for chrom in expected_chroms:
+    for region in expected_regions:
         # probably generate one big DataFrame with clustering
```
Comment (agalitsyna) [inline cooltools/dotfinder.py:1455]:
```suggestion
        # Perform clustering for each region separately.
```

File: cooltools/dotfinder.py [hunk 11]
```diff
@@ -1741,21 +1439,25 @@ def clustering_step(
...
         # information only and then just merge it with the
```
Comment (agalitsyna) [inline cooltools/dotfinder.py:1456]:
```suggestion
```

File: cooltools/dotfinder.py [hunk 12]
```diff
@@ -1741,21 +1439,25 @@ def clustering_step(
...
         # existing 'scores_df'-DataFrame.
```
Comment (agalitsyna) [inline cooltools/dotfinder.py:1457]:
```suggestion
```

File: cooltools/dotfinder.py [hunk 13]
```diff
@@ -1741,21 +1439,25 @@ def clustering_step(
...
         # should we use groupby instead of 'scores_df['chrom12']==chrom' ?!
```
Comment (agalitsyna) [inline cooltools/dotfinder.py:1458]:
```suggestion
```

File: cooltools/dotfinder.py [hunk 14]
```diff
@@ -1741,21 +1439,25 @@ def clustering_step(
...
         # to be tested ...
```
Comment (agalitsyna) [inline cooltools/dotfinder.py:1459]:
```suggestion
```

**Review by agalitsyna — CHANGES_REQUESTED** [review]
Docstring modifications

**Pull request merged**